src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIModelMapper.ChatCompletion.cs

```diff
@@ -0,0 +1,616 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Shared.Diagnostics;
+using OpenAI.Chat;
+
+#pragma warning disable SA1204 // Static elements should appear before instance elements
+#pragma warning disable S103 // Lines should not be too long
+#pragma warning disable CA1859 // Use concrete types when possible for improved performance
+
+namespace Microsoft.Extensions.AI;
+
+internal static partial class OpenAIModelMappers
+{
+    private static readonly JsonElement _defaultParameterSchema = JsonDocument.Parse("{}").RootElement;
+
+    public static OpenAI.Chat.ChatCompletion ToOpenAIChatCompletion(ChatCompletion chatCompletion, JsonSerializerOptions options)
+    {
+        _ = Throw.IfNull(chatCompletion);
+
+        if (chatCompletion.Choices.Count > 1)
+        {
+            throw new NotSupportedException("Creating OpenAI ChatCompletion models with multiple choices is currently not supported.");
+        }
+
+        List<OpenAI.Chat.ChatToolCall>? toolCalls = null;
+        foreach (AIContent content in chatCompletion.Message.Contents)
+        {
+            if (content is FunctionCallContent callRequest)
+            {
+                toolCalls ??= [];
+                toolCalls.Add(ChatToolCall.CreateFunctionToolCall(
+                    callRequest.CallId,
+                    callRequest.Name,
+                    new(JsonSerializer.SerializeToUtf8Bytes(
+                        callRequest.Arguments,
+                        options.GetTypeInfo(typeof(IDictionary<string, object?>))))));
+            }
+        }
+
+        OpenAI.Chat.ChatTokenUsage? chatTokenUsage = null;
+        if (chatCompletion.Usage is UsageDetails usageDetails)
+        {
+            chatTokenUsage = ToOpenAIUsage(usageDetails);
+        }
+
+        return OpenAIChatModelFactory.ChatCompletion(
+            id: chatCompletion.CompletionId,
+            model: chatCompletion.ModelId,
+            createdAt: chatCompletion.CreatedAt ?? default,
+            role: ToOpenAIChatRole(chatCompletion.Message.Role).Value,
```

Comment:
Seems like `ToOpenAIChatRole(chatCompletion.Message.Role)` could be null

Reply:
It's annotated `[return: NotNullIfNotNull("role")]` which is why the compiler isn't complaining about this.